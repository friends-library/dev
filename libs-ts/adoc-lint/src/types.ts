import { Lang, EditionType } from '@friends-library/types';

export interface LintResult {
  line: number;
  column: number | false;
  type: 'error' | 'warning' | 'notice';
  rule: string;
  message: string;
  recommendation?: string;
  fixable?: boolean;
  info?: string;
}

export interface LintOptions {
  lang: Lang;
  editionType?: EditionType;
  include?: string[];
  exclude?: string[];
  maybe?: boolean;
}

export interface LineRule {
  (line: string, lines: string[], lineNumber: number, options: LintOptions): LintResult[];
  slug: string;
  maybe?: boolean;
}

export interface BlockRule {
  (block: string, options: LintOptions): LintResult[];
  slug: string;
  maybe?: boolean;
}
