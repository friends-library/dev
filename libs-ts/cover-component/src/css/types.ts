export interface CoverCssModule {
  (scaler?: number, scope?: string): [string, string];
}
