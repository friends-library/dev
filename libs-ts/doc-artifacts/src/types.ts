export interface FileManifest {
  [key: string]: string | Buffer;
}

export interface Options {
  namespace?: string;
  srcPath?: string;
}

export type PdfOptions = Options & {
  formatOutput?: (line: string) => string;
};

export type EbookOptions = Options & {
  check: boolean;
};
