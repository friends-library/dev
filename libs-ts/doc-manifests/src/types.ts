import { PrintSize } from '@friends-library/types';

export interface PaperbackInteriorConfig {
  printSize: PrintSize;
  frontmatter: boolean;
  condense: boolean;
  allowSplits: boolean;
}

export interface PaperbackCoverConfig {
  printSize: PrintSize;
  volumes: number[];
  showGuides?: boolean;
}

export interface EbookConfig {
  frontmatter: boolean;
  subType: 'epub' | 'mobi';
  coverImg?: Buffer;
  randomizeForLocalTesting?: boolean;
}
