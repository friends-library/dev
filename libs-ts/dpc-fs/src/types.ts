import { DocPrecursor } from '@friends-library/types';

export type FsDocPrecursor = DocPrecursor & { fullPath: string };
