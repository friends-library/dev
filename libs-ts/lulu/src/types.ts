import { PrintSize } from '@friends-library/types';

export interface PrintSizeDetails {
  abbrev: PrintSize;
  maxPages: number;
  minPages: number;
  luluName: 'Pocket Book' | 'Digest' | 'US Trade';
  dims: {
    height: number;
    width: number;
  };
  margins: {
    top: number;
    bottom: number;
    outer: number;
    inner: number;
    runningHeadTop: number;
  };
}
