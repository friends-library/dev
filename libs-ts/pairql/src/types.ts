export type Env = 'dev' | 'staging' | 'prod';
export type Domain = 'dev' | 'evans-build' | 'order' | 'admin';
