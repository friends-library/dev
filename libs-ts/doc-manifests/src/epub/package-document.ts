import moment from 'moment';
import { DocPrecursor } from '@friends-library/types';
import { EbookSrcResult } from '@friends-library/evaluator';
import { t, setLocale } from '@friends-library/locale';
import ebookFrontmatter from './frontmatter';
import { EbookConfig } from '../types';

export function packageDocument(
  dpc: DocPrecursor,
  conf: EbookConfig,
  src: EbookSrcResult,
): string {
  const {
    lang,
    isCompilation,
    revision: { timestamp },
    meta: {
      author: { name, nameSort },
      title,
    },
  } = dpc;
  setLocale(lang);
  const modified = moment.utc(moment.unix(timestamp)).format(`YYYY-MM-DDThh:mm:ss[Z]`);
  const randomize = conf.randomizeForLocalTesting === true;
  const randomizer = ` (${moment().format(`h:mm:ss`)})`;
  const publisher = t`The Friends Library`;

  return `
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://www.idpf.org/2007/opf" version="3.0" unique-identifier="pub-id">
<metadata xmlns:dc="http://purl.org/dc/elements/1.1/">
  <dc:language id="pub-language">${lang}</dc:language>
  <dc:identifier id="pub-id">friends-library/${
    randomize ? Date.now() : dpc.editionId
  }</dc:identifier>
  <dc:title id="pub-title">${title}${randomize ? randomizer : ``}</dc:title>
  <dc:creator id="author">${isCompilation ? publisher : name}</dc:creator>
  <dc:publisher>${publisher}</dc:publisher>
  <dc:subject>${t`Quakers`}</dc:subject>
  <dc:subject>${t`Religious Society of Friends`}</dc:subject>
  <dc:rights>${t`Public domain in the USA`}.</dc:rights>
  ${isCompilation ? `` : `<meta property="file-as" refines="#author">${nameSort}</meta>`}
  <meta property="dcterms:modified">${modified}</meta>
  ${conf.coverImg ? `<meta name="cover" content="cover-img" />` : ``}
</metadata>
<manifest>
  ${[...manifestItems(dpc, conf, src)]
    .map(([id, data]) => `<item id="${id}" ${attrs(data)}/>`)
    .join(`\n  `)}
</manifest>
<spine>
  ${spineItems(dpc, conf, src)
    .map((id) => `<itemref idref="${id}"/>`)
    .join(`\n  `)}
</spine>
</package>
`.trim();
}

function attrs(data: Record<string, any>): string {
  return Object.entries(data)
    .reduce((acc, [key, val]) => {
      acc.push(`${key}="${String(val)}"`);
      return acc;
    }, [] as string[])
    .join(` `);
}

interface Item {
  href: string;
  'media-type': 'text/css' | 'application/xhtml+xml' | 'image/png' | 'image/jpeg';
  properties?: string;
}

export function manifestItems(
  dpc: DocPrecursor,
  conf: EbookConfig,
  src: EbookSrcResult,
): Map<string, Item> {
  const items = new Map<string, Item>();

  items.set(`css`, {
    href: `style.css`,
    'media-type': `text/css`,
  });

  if (conf.coverImg) {
    items.set(`cover-img`, {
      href: `cover.png`,
      'media-type': `image/png`,
      properties: `cover-image`,
    });

    items.set(`cover`, {
      href: `cover.xhtml`,
      'media-type': `application/xhtml+xml`,
    });
  }

  items.set(`nav`, {
    href: `nav.xhtml`,
    'media-type': `application/xhtml+xml`,
    properties: `nav`,
  });

  for (let num = 1; num <= src.numChapters; num++) {
    const id = `chapter-${num}`;
    items.set(id, {
      href: `${id}.xhtml`,
      'media-type': `application/xhtml+xml`,
    });
  }

  if (src.hasFootnotes) {
    items.set(`notes`, {
      href: `notes.xhtml`,
      'media-type': `application/xhtml+xml`,
    });
  }

  Object.keys(ebookFrontmatter(dpc, src, conf.subType)).forEach((slug) =>
    items.set(slug, {
      href: `${slug}.xhtml`,
      'media-type': `application/xhtml+xml`,
    }),
  );

  return items;
}

export function spineItems(
  dpc: DocPrecursor,
  conf: EbookConfig,
  src: EbookSrcResult,
): string[] {
  let items = Object.keys(ebookFrontmatter(dpc, src, conf.subType));
  items = items.concat(
    src.chapters.map((_, idx) => {
      return `chapter-${idx + 1}`;
    }),
  );

  if (src.hasFootnotes) {
    items.push(`notes`);
  }

  if (conf.coverImg) {
    items.unshift(`cover`);
  }

  return items;
}
