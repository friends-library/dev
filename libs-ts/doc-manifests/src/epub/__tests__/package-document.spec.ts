import { describe, it, expect } from '@jest/globals';
import { DocPrecursor, genericDpc, Lang } from '@friends-library/types';
import { evaluate } from '@friends-library/evaluator';
import { packageDocument } from '../package-document';

describe(`packageDocument()`, () => {
  it(`outputs correct string (English)`, () => {
    const dpc = dpcFromAdoc(`== Chapter 1\n\nHello world.\n`);
    const src = evaluate.toEbookSrcHtml(dpc);
    const conf = { frontmatter: true, subType: `epub` } as const;
    expect(packageDocument(dpc, conf, src)).toMatchSnapshot();
  });

  it(`outputs correct string (Spanish)`, () => {
    const dpc = dpcFromAdoc(`== Capitulo 1\n\nHola world.\n`, `es`);
    const src = evaluate.toEbookSrcHtml(dpc);
    const conf = { frontmatter: true, subType: `epub` } as const;
    expect(packageDocument(dpc, conf, src)).toMatchSnapshot();
  });
});

function dpcFromAdoc(adoc: string, lang: Lang = `en`): DocPrecursor {
  const dpc = genericDpc();
  dpc.lang = lang;
  dpc.asciidocFiles = [{ adoc, filename: `01-journal.adoc` }];
  return dpc;
}
