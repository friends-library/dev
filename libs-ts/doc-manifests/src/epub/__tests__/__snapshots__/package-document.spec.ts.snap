// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packageDocument() outputs correct string (English) 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"pub-id\\">
<metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
  <dc:language id=\\"pub-language\\">en</dc:language>
  <dc:identifier id=\\"pub-id\\">friends-library/9414033c-4b70-4b4b-8e48-fec037822173</dc:identifier>
  <dc:title id=\\"pub-title\\">Journal of George Fox</dc:title>
  <dc:creator id=\\"author\\">George Fox</dc:creator>
  <dc:publisher>The Friends Library</dc:publisher>
  <dc:subject>Quakers</dc:subject>
  <dc:subject>Religious Society of Friends</dc:subject>
  <dc:rights>Public domain in the USA.</dc:rights>
  <meta property=\\"file-as\\" refines=\\"#author\\">Fox, George</meta>
  <meta property=\\"dcterms:modified\\">2021-01-22T08:00:25Z</meta>
  
</metadata>
<manifest>
  <item id=\\"css\\" href=\\"style.css\\" media-type=\\"text/css\\"/>
  <item id=\\"nav\\" href=\\"nav.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav\\"/>
  <item id=\\"chapter-1\\" href=\\"chapter-1.xhtml\\" media-type=\\"application/xhtml+xml\\"/>
  <item id=\\"half-title\\" href=\\"half-title.xhtml\\" media-type=\\"application/xhtml+xml\\"/>
  <item id=\\"copyright\\" href=\\"copyright.xhtml\\" media-type=\\"application/xhtml+xml\\"/>
</manifest>
<spine>
  <itemref idref=\\"half-title\\"/>
  <itemref idref=\\"copyright\\"/>
  <itemref idref=\\"chapter-1\\"/>
</spine>
</package>"
`;

exports[`packageDocument() outputs correct string (Spanish) 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<package xmlns=\\"http://www.idpf.org/2007/opf\\" version=\\"3.0\\" unique-identifier=\\"pub-id\\">
<metadata xmlns:dc=\\"http://purl.org/dc/elements/1.1/\\">
  <dc:language id=\\"pub-language\\">es</dc:language>
  <dc:identifier id=\\"pub-id\\">friends-library/9414033c-4b70-4b4b-8e48-fec037822173</dc:identifier>
  <dc:title id=\\"pub-title\\">Journal of George Fox</dc:title>
  <dc:creator id=\\"author\\">George Fox</dc:creator>
  <dc:publisher>La Biblioteca de los Amigos</dc:publisher>
  <dc:subject>Cuáqueros</dc:subject>
  <dc:subject>La Sociedad de Amigos (Cuáqueros)</dc:subject>
  <dc:rights>Dominio público en los Estados Unidos de América.</dc:rights>
  <meta property=\\"file-as\\" refines=\\"#author\\">Fox, George</meta>
  <meta property=\\"dcterms:modified\\">2021-01-22T08:00:25Z</meta>
  
</metadata>
<manifest>
  <item id=\\"css\\" href=\\"style.css\\" media-type=\\"text/css\\"/>
  <item id=\\"nav\\" href=\\"nav.xhtml\\" media-type=\\"application/xhtml+xml\\" properties=\\"nav\\"/>
  <item id=\\"chapter-1\\" href=\\"chapter-1.xhtml\\" media-type=\\"application/xhtml+xml\\"/>
  <item id=\\"half-title\\" href=\\"half-title.xhtml\\" media-type=\\"application/xhtml+xml\\"/>
  <item id=\\"copyright\\" href=\\"copyright.xhtml\\" media-type=\\"application/xhtml+xml\\"/>
</manifest>
<spine>
  <itemref idref=\\"half-title\\"/>
  <itemref idref=\\"copyright\\"/>
  <itemref idref=\\"chapter-1\\"/>
</spine>
</package>"
`;
